// @linaria/utils/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  addIdentifierToLinariaPreval: true,
  getOrAddLinariaPreval: true,
  asyncResolveFallback: true,
  syncResolve: true,
  collectExportsAndImports: true,
  explicitImport: true,
  sideEffectImport: true,
  collectTemplateDependencies: true,
  extractExpression: true,
  createId: true,
  createPerfMeter: true,
  EventEmitter: true,
  findIdentifiers: true,
  nonType: true,
  findPackageJSON: true,
  hasEvaluatorMetadata: true,
  getFileIdx: true,
  hasMeta: true,
  getSource: true,
  isBoxedPrimitive: true,
  isExports: true,
  isNotNull: true,
  isRemoved: true,
  isRequire: true,
  isSerializable: true,
  isTypedNode: true,
  isUnnecessaryReactCall: true,
  removeDangerousCode: true,
  applyAction: true,
  mutate: true,
  removeWithRelated: true,
  findActionForNode: true,
  dereference: true,
  reference: true,
  referenceAll: true,
  slugify: true,
  ValueType: true,
  valueToLiteral: true,
  JSXElementsRemover: true
};
Object.defineProperty(exports, "EventEmitter", {
  enumerable: true,
  get: function () {
    return _EventEmitter.EventEmitter;
  }
});
Object.defineProperty(exports, "JSXElementsRemover", {
  enumerable: true,
  get: function () {
    return _JSXElementsRemover.default;
  }
});
Object.defineProperty(exports, "ValueType", {
  enumerable: true,
  get: function () {
    return _types.ValueType;
  }
});
Object.defineProperty(exports, "addIdentifierToLinariaPreval", {
  enumerable: true,
  get: function () {
    return _addIdentifierToLinariaPreval.addIdentifierToLinariaPreval;
  }
});
Object.defineProperty(exports, "applyAction", {
  enumerable: true,
  get: function () {
    return _scopeHelpers.applyAction;
  }
});
Object.defineProperty(exports, "asyncResolveFallback", {
  enumerable: true,
  get: function () {
    return _asyncResolveFallback.default;
  }
});
Object.defineProperty(exports, "collectExportsAndImports", {
  enumerable: true,
  get: function () {
    return _collectExportsAndImports.collectExportsAndImports;
  }
});
Object.defineProperty(exports, "collectTemplateDependencies", {
  enumerable: true,
  get: function () {
    return _collectTemplateDependencies.collectTemplateDependencies;
  }
});
Object.defineProperty(exports, "createId", {
  enumerable: true,
  get: function () {
    return _createId.createId;
  }
});
Object.defineProperty(exports, "createPerfMeter", {
  enumerable: true,
  get: function () {
    return _perfMetter.createPerfMeter;
  }
});
Object.defineProperty(exports, "dereference", {
  enumerable: true,
  get: function () {
    return _scopeHelpers.dereference;
  }
});
Object.defineProperty(exports, "explicitImport", {
  enumerable: true,
  get: function () {
    return _collectExportsAndImports.explicitImport;
  }
});
Object.defineProperty(exports, "extractExpression", {
  enumerable: true,
  get: function () {
    return _collectTemplateDependencies.extractExpression;
  }
});
Object.defineProperty(exports, "findActionForNode", {
  enumerable: true,
  get: function () {
    return _scopeHelpers.findActionForNode;
  }
});
Object.defineProperty(exports, "findIdentifiers", {
  enumerable: true,
  get: function () {
    return _findIdentifiers.default;
  }
});
Object.defineProperty(exports, "findPackageJSON", {
  enumerable: true,
  get: function () {
    return _findPackageJSON.findPackageJSON;
  }
});
Object.defineProperty(exports, "getFileIdx", {
  enumerable: true,
  get: function () {
    return _getFileIdx.default;
  }
});
Object.defineProperty(exports, "getOrAddLinariaPreval", {
  enumerable: true,
  get: function () {
    return _addIdentifierToLinariaPreval.getOrAddLinariaPreval;
  }
});
Object.defineProperty(exports, "getSource", {
  enumerable: true,
  get: function () {
    return _getSource.getSource;
  }
});
Object.defineProperty(exports, "hasEvaluatorMetadata", {
  enumerable: true,
  get: function () {
    return _hasEvaluatorMetadata.hasEvaluatorMetadata;
  }
});
Object.defineProperty(exports, "hasMeta", {
  enumerable: true,
  get: function () {
    return _hasMeta.hasMeta;
  }
});
Object.defineProperty(exports, "isBoxedPrimitive", {
  enumerable: true,
  get: function () {
    return _isBoxedPrimitive.isBoxedPrimitive;
  }
});
Object.defineProperty(exports, "isExports", {
  enumerable: true,
  get: function () {
    return _isExports.default;
  }
});
Object.defineProperty(exports, "isNotNull", {
  enumerable: true,
  get: function () {
    return _isNotNull.default;
  }
});
Object.defineProperty(exports, "isRemoved", {
  enumerable: true,
  get: function () {
    return _isRemoved.default;
  }
});
Object.defineProperty(exports, "isRequire", {
  enumerable: true,
  get: function () {
    return _isRequire.default;
  }
});
Object.defineProperty(exports, "isSerializable", {
  enumerable: true,
  get: function () {
    return _isSerializable.isSerializable;
  }
});
Object.defineProperty(exports, "isTypedNode", {
  enumerable: true,
  get: function () {
    return _isTypedNode.default;
  }
});
Object.defineProperty(exports, "isUnnecessaryReactCall", {
  enumerable: true,
  get: function () {
    return _isUnnecessaryReactCall.default;
  }
});
Object.defineProperty(exports, "mutate", {
  enumerable: true,
  get: function () {
    return _scopeHelpers.mutate;
  }
});
Object.defineProperty(exports, "nonType", {
  enumerable: true,
  get: function () {
    return _findIdentifiers.nonType;
  }
});
Object.defineProperty(exports, "reference", {
  enumerable: true,
  get: function () {
    return _scopeHelpers.reference;
  }
});
Object.defineProperty(exports, "referenceAll", {
  enumerable: true,
  get: function () {
    return _scopeHelpers.referenceAll;
  }
});
Object.defineProperty(exports, "removeDangerousCode", {
  enumerable: true,
  get: function () {
    return _removeDangerousCode.removeDangerousCode;
  }
});
Object.defineProperty(exports, "removeWithRelated", {
  enumerable: true,
  get: function () {
    return _scopeHelpers.removeWithRelated;
  }
});
Object.defineProperty(exports, "sideEffectImport", {
  enumerable: true,
  get: function () {
    return _collectExportsAndImports.sideEffectImport;
  }
});
Object.defineProperty(exports, "slugify", {
  enumerable: true,
  get: function () {
    return _slugify.default;
  }
});
Object.defineProperty(exports, "syncResolve", {
  enumerable: true,
  get: function () {
    return _asyncResolveFallback.syncResolve;
  }
});
Object.defineProperty(exports, "valueToLiteral", {
  enumerable: true,
  get: function () {
    return _valueToLiteral.valueToLiteral;
  }
});
var _addIdentifierToLinariaPreval = require("./addIdentifierToLinariaPreval");
var _asyncResolveFallback = _interopRequireWildcard(require("./asyncResolveFallback"));
var _collectExportsAndImports = require("./collectExportsAndImports");
var _collectTemplateDependencies = require("./collectTemplateDependencies");
var _createId = require("./createId");
var _perfMetter = require("./debug/perfMetter");
var _EventEmitter = require("./EventEmitter");
var _findIdentifiers = _interopRequireWildcard(require("./findIdentifiers"));
var _findPackageJSON = require("./findPackageJSON");
var _hasEvaluatorMetadata = require("./hasEvaluatorMetadata");
var _getFileIdx = _interopRequireDefault(require("./getFileIdx"));
var _hasMeta = require("./hasMeta");
var _getSource = require("./getSource");
var _isBoxedPrimitive = require("./isBoxedPrimitive");
var _isExports = _interopRequireDefault(require("./isExports"));
var _isNotNull = _interopRequireDefault(require("./isNotNull"));
var _isRemoved = _interopRequireDefault(require("./isRemoved"));
var _isRequire = _interopRequireDefault(require("./isRequire"));
var _isSerializable = require("./isSerializable");
var _isTypedNode = _interopRequireDefault(require("./isTypedNode"));
var _isUnnecessaryReactCall = _interopRequireDefault(require("./isUnnecessaryReactCall"));
var _options = require("./options");
Object.keys(_options).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _options[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _options[key];
    }
  });
});
var _removeDangerousCode = require("./removeDangerousCode");
var _scopeHelpers = require("./scopeHelpers");
var _slugify = _interopRequireDefault(require("./slugify"));
var _types = require("./types");
var _valueToLiteral = require("./valueToLiteral");
var _JSXElementsRemover = _interopRequireDefault(require("./visitors/JSXElementsRemover"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
